### PR TITLE
#701 Duplicated PMD and Checkstyle checks for multiple string literals

### PR DESCRIPTION
**Pull request by dskalenko**
Duplicated PMD and Checkstyle checks for multiple string literals #701

### CODE REVIEW

**Comment by davvd** [comment]
@dskalenko Let me find a reviewer for this pull request, thanks for submitting it

**Comment by davvd** [comment]
@mkordas it's yours,please review

**Comment by mkordas** [comment]
@dskalenko I'm on it

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/invoker.properties [hunk 1]
```diff
@@ -0,0 +1,2 @@
+invoker.goals = clean verify
+invoker.buildResult = failure
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/invoker.properties]: @dskalenko minor: newline missing

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @version $Id$
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml]: @dskalenko what's that `@version` here?

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>com.qulice.plugin</groupId>
+    <artifactId>pmd-duplicate-string-literals-violations</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <packaging>jar</packaging>
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml]: @dskalenko `jar` is default packaging

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+public final class FourDuplicateStringLiterals {
+
+    /**
+     * Construcor.
+     */
+    private FourDuplicateStringLiterals() { }
+
+    /**
+     * Method with four
+     * duplicate string literals.
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java]: @dskalenko join lines

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * @author Denys Skalenko (d.skalenko@gmail.com)
+ * @version $Id$
+ * @since 1.0
+ */
+public final class FourDuplicateStringLiterals {
+
+    /**
+     * Construcor.
+     */
+    private FourDuplicateStringLiterals() { }
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java]: @dskalenko do we need constructor in this input?

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.qulice.foo;
+
+/**
+ * Class with four duplicate string literals.
+ * @author Denys Skalenko (d.skalenko@gmail.com)
+ * @version $Id$
+ * @since 1.0
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java]: @dskalenko let's fill it with correct value

File: qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml [hunk 7]
```diff
@@ -61,7 +61,9 @@
         <exclude name="GuardLogStatementJavaUtil"/>
     </rule>
     <rule ref="rulesets/java/optimizations.xml"/>
-    <rule ref="rulesets/java/strings.xml"/>
+    <rule ref="rulesets/java/strings.xml">
+        <exclude name="AvoidDuplicateLiterals"/>
```
Comment (mkordas) [inline qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml]: @dskalenko what's wrong with original rule? why you excluded it?

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/TwoDuplicateStringLiterals.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+     * Construcor.
+     */
+    private TwoDuplicateStringLiterals() { }
+
+    /**
+     * Method with two duplicate string literals.
+     * @return Formated string
+     */
+    public static String methodWithTwoDuplicateStringLiterals() {
+        return String.format("%s%s", "test", "test");
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/TwoDuplicateStringLiterals.java]: @dskalenko can we also add test case for suppression?

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/WithoutDuplicateStringLiterals.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.qulice.foo;
+
+/**
+ * Class without duplicate string literals.
+ * @author Denys Skalenko (d.skalenko@gmail.com)
+ * @version $Id$
+ * @since 1.0
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/WithoutDuplicateStringLiterals.java]: @dskalenko should be `0.17`

File: qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml [hunk 10]
```diff
@@ -196,4 +198,17 @@
...
+    <rule name="AvoidDuplicateLiterals"
+          message="The String literal {0} appears {1} times in this file; the first occurrence is on line {2}"
+          class="net.sourceforge.pmd.lang.java.rule.strings.AvoidDuplicateLiteralsRule">
+        <description>
+            Code containing duplicate String literals can usually be improved by declaring the String as a constant field.
+        </description>
+        <priority>3</priority>
+        <properties>
+            <property name="maxDuplicateLiterals">
+                <value>2</value>
```
Comment (mkordas) [inline qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml]: @dskalenko does it mean that we allow 2 duplicates?

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    private FourDuplicateStringLiterals() { }
+
+    /**
+     * Method with four
+     * duplicate string literals.
+     * @return Formated string
+     */
+    public static String methodWithFourDuplicateStringLiterals() {
+        return String.format("%s%s%s%s", "test", "test", "test", "test");
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java]: @dskalenko please also add test case for just 2 duplicates

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * Validate that the build really failed and violations were reported
+ * because of PMD AvoidDuplicateLiterals.
+ * @version $Id$
+ */
+def log = new File(basedir, 'build.log')
+assert !log.text.contains('MultipleStringLiteralsCheck')
+assert !log.text.contains('WithoutDuplicateStringLiterals.java')
+assert log.text.contains('FourDuplicateStringLiterals.java[51-51]: The String ' +
+        'literal "test" appears 4 times in this file; the first occurrence is ' +
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy]: @dskalenko should be indented with just 4 spaces

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Validate that the build really failed and violations were reported
+ * because of PMD AvoidDuplicateLiterals.
+ * @version $Id$
+ */
+def log = new File(basedir, 'build.log')
+assert !log.text.contains('MultipleStringLiteralsCheck')
+assert !log.text.contains('WithoutDuplicateStringLiterals.java')
+assert log.text.contains('FourDuplicateStringLiterals.java[51-51]: The String ' +
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy]: @dskalenko `+` is traditionally on new line

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Validate that the build really failed and violations were reported
+ * because of PMD AvoidDuplicateLiterals.
+ * @version $Id$
+ */
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy]: @dskalenko missing since

**Comment by mkordas** [comment]
@dskalenko please check why we have merge conflict

File: qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml [hunk 15]
```diff
@@ -196,4 +198,17 @@
...
     </rule>
+    <rule name="AvoidDuplicateLiterals"
+          message="The String literal {0} appears {1} times in this file; the first occurrence is on line {2}"
+          class="net.sourceforge.pmd.lang.java.rule.strings.AvoidDuplicateLiteralsRule">
+        <description>
+            Code containing duplicate String literals can usually be improved by declaring the String as a constant field.
+        </description>
+        <priority>3</priority>
+        <properties>
+            <property name="maxDuplicateLiterals">
```
Comment (mkordas) [inline qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml]: @dskalenko I think it's enough to just modify this property, no need to redefine the check

**Comment by mkordas** [comment]
@dskalenko please check my comments

**Comment by mkordas** [comment]
@dskalenko ping

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml (hunk 2 above)
Comment (dskalenko) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml]: @mkordas I'll remove it

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml (hunk 3 above)
Comment (dskalenko) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml]: @mkordas I know :) but what is wrong here? :)

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java (hunk 5 above)
Comment (dskalenko) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java]: @mkordas yes we need it, for avoiding other violation

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java (hunk 6 above)
Comment (dskalenko) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java]: @mkordas ok I'll fix it

File: qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml (hunk 10 above)
Comment (dskalenko) [inline qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml]: @mkordas no we don't allowed 2 duplicates, 2 or more it's PMD violation

File: qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml (hunk 7 above)
Comment (dskalenko) [inline qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml]: @mkordas I excluded - to be able add with custom configuration

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/WithoutDuplicateStringLiterals.java (hunk 9 above)
Comment (dskalenko) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/WithoutDuplicateStringLiterals.java]: @mkordas ok

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy (hunk 14 above)
Comment (dskalenko) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy]: @mkordas ok I'll add

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/TwoDuplicateStringLiterals.java (hunk 8 above)
Comment (dskalenko) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/TwoDuplicateStringLiterals.java]: @mkordas suppression for PMD AvoidDuplicateLiterals?

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml (hunk 3 above)
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml]: @dskalenko nothing wrong, but if it's default, let's remove it

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java (hunk 5 above)
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java]: @dskalenko OK, thanks

File: qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml (hunk 7 above)
Comment (mkordas) [inline qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml]: @dskalenko you don't need to exclude to provide custom configuration

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/TwoDuplicateStringLiterals.java (hunk 8 above)
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/TwoDuplicateStringLiterals.java]: @dskalenko yes - test confirming that only one suppression is enough to suppress - previously we needed two

File: qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml (hunk 10 above)
Comment (mkordas) [inline qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml]: @dskalenko OK, I see

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy (hunk 14 above)
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy]: @dskalenko thanks

File: qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml (hunk 15 above)
Comment (dskalenko) [inline qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml]: @mkordas Agree - will fix it

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/TwoDuplicateStringLiterals.java (hunk 8 above)
Comment (dskalenko) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/TwoDuplicateStringLiterals.java]: @mkordas  ok I'll add test for this case

File: qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml (hunk 7 above)
Comment (dskalenko) [inline qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml]: @mkordas yes I'll fix it

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml (hunk 3 above)
Comment (dskalenko) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/pom.xml]: @mkordas ok

**Comment by dskalenko** [comment]
@mkordas thanks for you review, please take a look my fix

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+ */
+package com.qulice.foo;
+
+/**
+ * Class with four duplicate string literals.
+ * @author Denys Skalenko (d.skalenko@gmail.com)
+ * @version $Id$
+ * @since 0.17
+ */
+@SuppressWarnings("PMD.ProhibitPublicStaticMethods")
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/FourDuplicateStringLiterals.java]: @dskalenko this suppression and private constructor and too noisy for me. Can we have an object without constructor and with one non-static method that would do the job? Method can remain non-static if it will refer to `this` at least once.

File: qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml [hunk 17]
```diff
@@ -62,6 +62,13 @@
     </rule>
     <rule ref="rulesets/java/optimizations.xml"/>
     <rule ref="rulesets/java/strings.xml"/>
+    <rule ref="rulesets/java/strings.xml/AvoidDuplicateLiterals">
+        <properties>
+            <property name="maxDuplicateLiterals">
+                <value>2</value>
```
Comment (mkordas) [inline qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml]: @dskalenko much better now :)

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Validate that the build really failed and violations were reported
+ * because of PMD AvoidDuplicateLiterals.
+ * @version $Id$
+ * @since 0.17
+ */
+def log = new File(basedir, 'build.log')
+assert !log.text.contains('MultipleStringLiteralsCheck')
+assert !log.text.contains('WithoutDuplicateStringLiterals.java')
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/verify.groovy]: @dskalenko shouldn't we add `SuppressDuplicateStringLiterals.java` as well?

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/SuppressDuplicateStringLiterals.java [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    private SuppressDuplicateStringLiterals() { }
+
+    /**
+     * Method with two duplicate string literals.
+     * @return Formated string
+     */
+    @SuppressWarnings("PMD.AvoidDuplicateLiterals")
+    public static String methodWithTwoDuplicateStringLiterals() {
+        return String.format("%s%s", "test", "test");
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/SuppressDuplicateStringLiterals.java]: @dskalenko I don't think that we need this test case

**Comment by mkordas** [comment]
@dskalenko please see my responses, generally PR is in good shape now

**Comment by mkordas** [comment]
@dskalenko ping

**Comment by mkordas** [comment]
@dskalenko can you continue here?

**Comment by dskalenko** [comment]
@mkordas could you please take a look now

File: qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/SuppressDuplicateStringLiterals.java [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+public final class SuppressDuplicateStringLiterals {
+
+    /**
+     * Method with four duplicate string literals.
+     * @return Formated string
+     */
+    @SuppressWarnings("PMD.AvoidDuplicateLiterals")
+    public String methodWithFourDuplicateStringLiterals() {
+        return String.format(
+            "%s%s%s%s%s", "test", "test", "test", "test", this.toString()
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-duplicate-string-literals-violations/src/main/java/com/qulice/foo/SuppressDuplicateStringLiterals.java]: @dskalenko looks much better now :)

**Comment by mkordas** [comment]
@dskalenko good work!

**Comment by mkordas** [comment]
@rultor merge

**Comment by krzyk** [comment]
@dskalenko please add reference to the issue in the description

**Comment by mkordas** [comment]
@krzyk I should have spotted that, thanks for being vigilant

**Comment by dskalenko** [comment]
#727

**Comment by krzyk** [comment]
@dskalenko please add it to the description

**Comment by dskalenko** [comment]
@krzyk done, sorry for that

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by krzyk** [comment]
@dskalenko you added reference to the PR, not the issue

**Comment by dskalenko** [comment]
@krzyk fixed

**Comment by krzyk** [comment]
@rultor deploy

**Comment by davvd** [comment]
@mkordas quality is good here, 10 mins added to @krzyk (our architect) in transaction 79073790; Many thanks! **50 mins** were added to your account in Transaction ID `56d5a15add127d7962000666` (task took 123 hours and 14 mins); you're getting extra minutes here (c=35); +50 added to your rating, at the moment it is: [+7815](http://www.netbout.com/b/36080?open=rating)

**Comment by davvd** [comment]
@rultor deploy now pls

**Comment by krzyk** [comment]
@davvd this is not urgent

**Comment by davvd** [comment]
> @davvd this is not urgent

@krzyk right, I removed "urgent" tag
